firedrake/slate/slac/utils.py

```diff
@@ -136,6 +152,224 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+# singledispatch for class functions
+def classsingledispatch(func):
+    dispatcher = singledispatch(func)
+
+    def wrapper(*args, **kw):
+        return dispatcher.dispatch(args[1].__class__)(*args, **kw)
+
+    wrapper.register = dispatcher.register
+    update_wrapper(wrapper, func)
+    return wrapper
```

Comment:
This would be unnecessary if `SlateTranslator.slate_to_gem_translate` were not on the class `SlateTranslator`.

Reply:
Yes, but then I would not be able to let the `Translator` own the `Builder` in form of class attribute. I could pass it to slate_to_gem_translate instead if that is better.

Reply:
I will explain this later...